### ISSUE TITLE
Loading Images on recycled GridView cells

### ISSUE THREAD

**Issue by cdep** [koush/UrlImageViewHelper#17]
I may be using it wrong, but images are appearing on recycled cells in my GridView, essentially making duplicates. Is this how it's supposed to work, or am I doing something wrong?

If I have 4 images in a 3-column GridView, I get:
    [image 1] [image 2] [image 3]
    [image 4] [image 2] [image 1]
    [image 1]

I read a bit more of your code, and added some logs to the part with the comment: "since listviews recycle their views, we need to take note of which url this view is waiting for". I'm assuming this doesn't take into consideration what would happen if the view is set to null?

**Comment by koush** [comment]
Got some simple sample code I can take a look at?

The recycle code is around views being recycled in adapters.

**Comment by cdep** [comment]
Right now, I'm working on this gallery project, and the focus is on the adapter. Would you like me to email it to you? Or should I just post my adapter's getView() method?

**Comment by koush** [comment]
Sure, if you don't mind posting the code, I'd be happy to take a look at it.

**Comment by koush** [comment]
k o u s h d _ at _ gmail

**Comment by cdep** [comment]
I added some comments so the code is more understandable. The bulk of the code is in com.example.PhotoListAdapter. Everything else is cookie-cutter.

http://dl.dropbox.com/u/11065606/sample-gallery.zip

**Comment by koush** [comment]
I see this:
http://download.clockworkmod.com/screenshots//Device_Screen_Capture-20120903-003237.png

That seems to be correct. You on the latest version of the library? What version of Android?

**Comment by koush** [comment]
I'd also recommend not using a view lookup table with any adapters. That usually is problematic. Try using getParent and findViewById and such to navigate the view hierarchy.

**Comment by cdep** [comment]
API 8. As for being on the latest version, I called 'git pull' and it said I was up-to-date. The squares are supposed to be uniform in height, rather than rectangles of varying heights. I supposedly got that one fixed last week, and was down to the images repeating. I guess I was wrong. 

As for the lookup table, I thought of using that to fix the duplicating problem i.e. whenever I'd call UrlImageViewHelper.setUrlDrawable..., I'd map the ViewSwitcher I used to the url, then check if the ImageView from the table was the same as the ImageView passed to the callback. I'd call imageView.setImageDrawable(null) if it wasn't. I think I made a mistake there cause I only checked if the mapping was null, rather than checking if the ImageViews were the same. In any case, I scrapped the mapping and tried out another approach.

**Comment by koush** [comment]
I got rid of all the cell mapping and just did this, and it worked fine:

``` java
    @Override
    public void onLoaded(final ImageView imageView,
            final Drawable loadedDrawable, final String url,
            final boolean loadedFromCache) {
        final ViewSwitcher cell = (ViewSwitcher)imageView.getParent();
      String msg = String.format("Displaying image %s on ViewSwitcher %s",url, cell.toString());
      showImageView(cell);
//        String msg;
//        if (cell != null) {
//            msg = String.format("Displaying image %s on ViewSwitcher %s",
//                    url, cell.toString());
//            showImageView(cell);
//        } else {
//            imageView.setImageDrawable(null);
//            msg = "ViewSwitcher not found";
//        }
//        Log.i(TAG + " onLoaded", msg);
    }
```

**Comment by koush** [comment]
Instead of faking out getCount(), I would maybe just populate the items in the list with nulls to make sure a row is filled out. I suspect messing with getCount can screw with things.

**Comment by cdep** [comment]
[Here's what I've been seeing.](http://dl.dropbox.com/u/11065606/device-2012-09-03-160321.png) I also tried removing the cell mappings, but got the same results. 

Oh, I managed to get it working by adding 

``` java
if (cell.getCurrentView() instanceof ImageView) {
    imageView.setImageDrawable(null);
}
```

I'll test it out with two screens worth of images to see if I hit any problems with recycled views that aren't null. UrlImageViewHelper loading of images into nulled recycled views could be a small issue to some people. I tried passing null to the loader e.g., 

```
UrlImageViewHelper.setImageDrawable(imageView, null, this);
```

but I still duplicates.

**Comment by koush** [comment]
You're running this on an API 8 emulator you said?

**Comment by koush** [comment]
When passing "null" to the loader, try not to pass the "this" for the onLoaded callback. see if that fixes it.

**Comment by cdep** [comment]
**Summary:** I cleaned up the code and it works better, but I think `UrlImageViewHelper` doesn't play nice with recycled views.

Yes, I'm testing on an API 8 emulator. I'm updating the code that I uploaded to dropbox. I cleaned it up a lot, so it should be easier to read. I also applied a lot of what you said: I stopped messing with getCount(), I put the `ImageViews` in your library's hands, i.e. I don't touch the `ImageViews` any more, and I removed the view mapping.

I'm still having problems though. Some of the cells keep coming out blank. The blank cell isn't constant, and changes when you scroll, so I assume it's the recycled view that's showing up blank, rather than an image that comes out null. I don't call `ImageView.setImageDrawable(null)` anymore, so I don't get why it's showing up empty.

Also, if you scroll around while the images are loading, some of the loaded images get replaced. My guess is this has something to do with `UrlImageViewHelper` calling `setImage` of all the `ImageViews` that were registered to load the newly-loaded image. This clears up after all the images are loaded and you have them reload by scrolling them out of view. I'll try turning view recycling off to see if that fixes it.
